### PR TITLE
Update Function Semantics

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation
The current update command for functions/sources/sinks expects us to supply all the details. This is non-untuitive and forms a very bad user experience since updates typically affect only a few parameters.
This pr re-organizes source/sink/function updates to allow only a subset of details. We do a merge with the existing config, flag any errors before updating.

### Modifications

Describe the modifications you've done.

### Result

After your change, what will change.

### CODE REVIEW

**Comment by srkukarni** [comment]
rerun integration tests

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java [hunk 1]
```diff
@@ -691,6 +701,23 @@ void runCmd() throws Exception {
 
     @Parameters(commandDescription = "Update a Pulsar Function that's been deployed to a Pulsar cluster")
     class UpdateFunction extends FunctionDetailsCommand {
+
+        protected void validateFunctionConfigs(FunctionConfig functionConfig) {
```
Comment (jerrypeng) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: Please add Override annotation

**Comment by jerrypeng** [comment]
@srkukarni this PR looks mostly good.  Though we probably want to think about how a restart operation is carried out.  Do we need to re-download the function package for every type of update or can we only do that if the package is updated? Do we need to tear down the entire stateful set in kubernetes for a restart.  Since a restart operation is just stop and start operation in the backend.

File: pulsar-broker/src/test/java/org/apache/pulsar/functions/worker/PulsarWorkerAssignmentTest.java [hunk 2]
```diff
@@ -264,7 +264,7 @@ public void testFunctionAssignmentsWithRestart() throws Exception {
                     "my.*", sinkTopic, subscriptionName);
             functionConfig.setParallelism(parallelism);
             // set-auto-ack prop =false
```
[the remark this answers is not in the document]
Comment (srkukarni) [reply pulsar-broker/src/test/java/org/apache/pulsar/functions/worker/PulsarWorkerAssignmentTest.java]: Updated

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java (hunk 1 above)
Comment (srkukarni) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: Added

**Comment by srkukarni** [comment]
@jerrypeng I've addressed your comments.
Note that this is pr primarily focuses on the functions front-end and already does some of the optimizations that you mentioned. For instance if there is no change in the function code, we no longer upload a new copy to bk(We do a download however from existing bk location just to make sure that all the parameters are ok, this cannot be avoided). 
Making restart a first class citizen in the runtime is a follow up change

**Pull request merged**